src/txmempool.cpp

```diff
@@ -1438,24 +1438,24 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint> *pvNoSpends
 
 void CTxMemPool::UpdateTransactionsPerSecond()
 {
-    boost::mutex::scoped_lock lock(cs_txPerSec);
-
+    static std::atomic<double> nTxnsProcessed{0};
     static int64_t nLastTime = GetTime();
+
     double nSecondsToAverage = 60; // Length of time in seconds to smooth the tx rate over
     int64_t nNow = GetTime();
 
-    // Decay the previous tx rate.
-    int64_t nDeltaTime = nNow - nLastTime;
-    if (nDeltaTime > 0)
+    double nNewTxns = 0.0;
+    double nOldTxns = 0.0;
+    do
     {
-        nTxPerSec -= (nTxPerSec / nSecondsToAverage) * nDeltaTime;
-        nLastTime = nNow;
-    }
-
-    // Add the new tx to the rate
-    nTxPerSec += 1 / nSecondsToAverage; // The amount that the new tx will add to the tx rate
-    if (nTxPerSec < 0)
-        nTxPerSec = 0;
+        // decay the number of transactions over nSecondsToAverage and then add "1" for
+        // the new transaction.
+        nOldTxns = nTxnsProcessed.load();
```

Comment:
Can't / shouldn't

`nOldTxns = nTxnsProcessed.load(); `

be pulled before the `do`-loop as it is updated in `compare_exchange_weak(..)`?

Reply:
if compare_exchange_weak fails the first time. we need the current value of nTxnsProcessed to do our calculation with since it could have been changed by another process.

Reply:
ah, yes nvm , you're right.